src/generic_datatype.cpp

```diff
@@ -0,0 +1,146 @@
+#include "generic_datatype.h"
+
+#include <algorithm>
+#include <exception>
+#include <memory>
+#include <unordered_map>
+
+#include "assert.h"
+
+using namespace std;
+
+namespace smt {
+
+GenericDatatypeDecl::GenericDatatypeDecl(const std::string name) : dt_name(name)
+{
+}
+
+std::string GenericDatatypeDecl::get_name() { return dt_name; }
+
+GenericDatatypeConstructorDecl::GenericDatatypeConstructorDecl(
+    const std::string & name)
+  : cons_name(name),
+    // Should I turn this into a constant?
+    dt_name("NOT IMPLEMENTED YET")
+{
+}
+
+void GenericDatatypeConstructorDecl::add_new_selector(
+    const SelectorComponents & newSelector)
+{
+  // This seems like an opportunity for a new comparison function for
+  // the struct.
+  for (unsigned int i = 0; i < selector_vector.size(); ++i)
+  {
+    if (selector_vector[i].name == (newSelector).name)
+    {
+      throw "Can't add selector. It already exists in vector!";
+    }
+  }
+  selector_vector.push_back(newSelector);
+}
+
+std::vector<SelectorComponents>
+GenericDatatypeConstructorDecl::get_selector_vector()
+{
+  return selector_vector;
+}
+
+std::string GenericDatatypeConstructorDecl::get_name() const
+{
+  return cons_name;
+}
+
+int GenericDatatypeConstructorDecl::get_selector_count() const
+{
+  return selector_vector.size();
+}
+bool GenericDatatypeConstructorDecl::compare(
+    const DatatypeConstructorDecl & d) const
+{
+  // Why won't type casting like this work?
+  // GenericDatatypeConstructorDecl gdtc = (GenericDatatypeConstructorDecl) d;
+  return cons_name
+         == static_pointer_cast<GenericDatatypeConstructorDecl>(d)->get_name();
+}
+
+  std::string GenericDatatypeConstructorDecl::get_dt_name() const
+  {
+    return dt_name;
+  }
+
+  void GenericDatatypeConstructorDecl::change_dt_name(const std::string new_name)
```

Comment:
This is a bit scary -- why would we need that?

Reply:
This was the first approach I thought of to give a value to a protected member after initialization. Because we only know the dt_name/datatype when we are actually adding a constructor to a datatype, I needed some way of updating the member. Would you recommend a particular approach to this task because this problem persists after transitioning over to using a datatype member instead.

Reply:
There is the name of the constructor, and the name of the datatype -- these are different -- don't you know the name of the datatype already when you create a `DatatypeDecl`?

Reply:
Yes we have access to the datatype name through each datatypedecl. My problem is to have a member variable in the constructor class that keeps track of the datatype/datatypedecl itself. As far as I can tell, we need a way to change the value of that member variable after initializaiton.

Reply:
But why would the datatype/datatypedecl's name will change after you add a constructor to it?

Reply:
I see now that the function name is misleading. I never intended to change the datatype's name with the function, merely the member called "dt_name" that the constructor class holds. This function was only intended to be called once when a constructor is added to a datatype so that the constructor has a record of the name of the datatype its associated with.

Reply:
Let's take this particular issue offline tomorrow :)

Reply:
I see now that the function name is misleading. I never intended to change the datatype's name with the function, merely the member called "dt_name" that the constructor class holds. This function was only intended to be called once when a constructor is added to a datatype so that the constructor has a record of the name of the datatype its associated with.